kombu/transport/sqlalchemy/__init__.py

```diff
@@ -130,10 +146,16 @@ def _size(self, queue):
         return self._query_all(queue).count()
 
     def _declarative_cls(self, name, base, ns):
-        with _MUTEX:
-            if name in class_registry:
-                return class_registry[name]
-            return type(str(name), (base, ModelBase), ns)
+        if name not in class_registry:
+            with _MUTEX:
+                if name in class_registry:
+                    # Class was registered while we were waiting to
+                    # acquire the lock.
+                    return class_registry[name]
+
+                return type(str(name), (base, ModelBase), ns)
```

Comment:
Even though this method was already fixed I tweaked it to only hold the mutex when strictly necessary, which will improve performance.

Reply:
can you update some related tests as well?

Reply:
Sure, can you let me know which tests you have in mind? I'm still finding my way around the project somewhat. Are you referring to `t/unit/transport/test_sqlalchemy.py`?

Reply:
yes

Reply:
Ok great, I'll add a few test cases later today and update the PR.

Reply:
looking forward to it.